### ISSUE TITLE
error[E0502] (cannot borrow with different mutability) in code permitted by Edition 2024

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust-clippy#15160]
### Summary

I'm not quite sure how to report this, since this appears to be a false positive in a `rustc` lint when using Clippy. Maybe it doesn't fully respect the lifetime changes in Edition 2024?

Under Edition 2021 I had the following code:
```rust
        let mut messages = MessageStack::new();
        let widget = self.widget.as_node(&state.data);

        if Some(requested_resume) == self.ev_state.next_resume() {
            self.ev_state
                .with(&mut state.shared, window, &mut messages, |cx| {
                    cx.update_timer(widget);
                });
        } else {
            drop(widget); // make the borrow checker happy
        }
```
With Edition 2024, the explicit `drop(widget)` is no longer necessary (`cargo check` passses), yet Clippy still complains:
```
error[E0502]: cannot borrow `*state` as mutable because it is also borrowed as immutable
   --> crates/kas-core/src/runner/window.rs:392:9
    |
383 |         let widget = self.widget.as_node(&state.data);
    |                                          ----------- immutable borrow occurs here
...
392 |         state.handle_messages(&mut messages);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
393 |     }
    |     - immutable borrow might be used here, when `widget` is dropped and runs the destructor for type `core::node::Node<'_>`
```

See https://github.com/kas-gui/kas/pull/514

Using: rustc 1.90.0-nightly (b03b3a7ec 2025-06-26)
Using: clippy 0.1.89 (b03b3a7ec9 2025-06-26)

**Comment by ComputerDruid** [comment]
I can't reproduce. I tried both a minimal example in the playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=735254a4e7871dc7c34ff91436af96f6 (which does not compile)

and I tried checking out the linked pull request (at 74bd999c631ac86ffc099d26567078a0e3675804), and commenting out the specified line, and getting the same error with `cargo check`:

```
error[E0502]: cannot borrow `*state` as mutable because it is also borrowed as immutable
   --> crates/kas-core/src/runner/window.rs:394:9
    |
383 |         let widget = self.widget.as_node(&state.data);
    |                                          ----------- immutable borrow occurs here
...
394 |         state.handle_messages(&mut messages);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
395 |     }
    |     - immutable borrow might be used here, when `widget` is dropped and runs the destructor for type `node::Node<'_>`

For more information about this error, try `rustc --explain E0502`.
error: could not compile `kas-core` (lib) due to 1 previous error
```

Which makes sense to me, I don't see why edition 2024 would change the drop order here.

**Comment by dhardy** [comment]
Sorry, I can't reproduce this either.
